src/PIL/Image.py

```diff
@@ -2436,16 +2584,23 @@ def fromarray(obj, mode=None):
         raise ValueError("Too many dimensions: %d > %d." % (ndim, ndmax))
 
     size = shape[1], shape[0]
+
+    obj_bytes = None # type: bytes
     if strides is not None:
         if hasattr(obj, 'tobytes'):
-            obj = obj.tobytes()
+            obj_bytes = obj.tobytes()
         else:
-            obj = obj.tostring()
+            obj_bytes = obj.tostring()
 
-    return frombuffer(mode, size, obj, "raw", rawmode, 0, 1)
+    # Typing: At this point, obj should be an item exporting a buffer interface
+    # and obj_bytes, if it exists, is a bytes exporting a buffer interface.
+    # Unfortunately at this point, there's no protocol for the buffer interface
+    # and due to it's c-level implementation, I don't see how to make one.
+    return frombuffer(mode, size, obj_bytes or obj, "raw", rawmode, 0, 1) # type: ignore
```

Comment:
Maybe file an issue in typeshed re this, if there's not one already? If typeshed is the best place - maybe even python/typing? (for more general language/typing queries)